Initial implementation

johnnychen94 opened the pull request:
Follow the proposal in https://github.com/JuliaGeometry/CoordinateTransformations.jl/issues/75#issuecomment-902692944, there are three core types here:

* `AbstractCoordinateSystem`: this defines how to generate the coordinate map from/to cartesian system
* `AbstractPoint`: every point lives in some space and thus has its associated coordinate system
* `AbstractCoordinateTransformation`: every coordinate transformation is a map from one point space to another point space

The most useful and efficient coordinate and point representation is perhaps the one that preserves the origin when maps to/from the euclidean space, i.e., the coordinate map that satisfies `ϕ(0) = 0`. I named the system `FixedOriginCoordinateSystem`, and the point type `Point`.

As a simple test, here's how cartesian translation is applied to point:

```julia
using CoordinateSystems

# by default it lives in Cartesian system
p = point(2, 3)
trans = Translation(1, 2)
trans(p) # Point(3, 5)

# doing translation on Cartesian system
p_s = point(Spherical{2}(), p)
q_s = point(Cartesian{2}(), trans(p_s))
q_s == trans(p) # true
```

As a benchmark to compare with CoordinateTransformations:

```julia
using StaticArrays
import CoordinateTransformations

trans_CT = CoordinateTransformations.Translation(1, 2)
p_CT = SVector(2, 3)
@btime $(Ref(trans_CT))[]($(Ref(p_CT))[]); # 1.983 ns (0 allocations: 0 bytes)
@btime $(Ref(trans))[]($(Ref(p))[]); # 1.982 ns (0 allocations: 0 bytes)
```

johnnychen94 committed:
Files generated by PkgTemplates

PkgTemplates version: 0.7.17

johnnychen94 committed:
WIP: initial implementation

johnnychen94 commented on src/transformations.jl:
```diff
@@ -0,0 +1,32 @@
+@inline Translation(x) = Translation(coordinate_system(x), x)
+@inline Translation(xs::Real...) = Translation(SVector(xs))
+
+function (trans::Translation)(x::AbstractPoint)
+    with_coordinate_system(trans.cs, x) do p
+        SVector(p) + trans.translation
+    end
+end
```
I'm not sure here if we want the output point live in the coordinate system of `trans` or live in the coordinate system of input `x`. The current implementation chooses the latter one but obviously, if we're having mixed coordinate systems during computation there will be performance overhead.

juliohm commented on src/CoordinateSystemsCore/CoordinateSystemsCore.jl:
```diff
@@ -0,0 +1,33 @@
+module CoordinateSystemsCore
```
Can you explain how we benefit from a Core submodule?

juliohm commented on src/CoordinateSystemsCore/CoordinateSystemsCore.jl:
```diff
@@ -0,0 +1,33 @@
+
+using StaticArrays
+
+export AbstractCoordinateSystem, Cartesian, Spherical
+export point, coordinate, coordinate_system, with_coordinate_system
+export AbstractPoint, Point
```
I would remove the Abstract prefix from most types to make them short. Also, maybe we could leave the Point type outside this package and only deal with the coordinate transformations. The point itself can be defined in separate packages that consume coordinates.

juliohm commented on src/CoordinateSystemsCore/CoordinateSystemsCore.jl:
```diff
@@ -0,0 +1,33 @@
+export AbstractPoint, Point
+
+# We accept several user input types, but always store and compute as SVector
+const CoordinateType{N} = Union{NTuple{N}, SVector{N}}
```
I would not fall back to SVector or NTuple because that is gonna cause the same issues we have currently with SVector everywhere. We need to really enforce some Cartesian type to avoid SVector proliferation.

johnnychen94 replied:
This would allow users to create points in a convenient way, e.g., `point(2, 3)`. But internally it's still stored and computed as `SVector`.

juliohm commented on src/CoordinateSystemsCore/CoordinateSystemsCore.jl:
```diff
@@ -0,0 +1,33 @@
+const CoordinateType{N} = Union{NTuple{N}, SVector{N}}
+
+abstract type AbstractCoordinateSystem{N} end
+abstract type FixedOriginCoordinateSystem{N} <: AbstractCoordinateSystem{N} end
```
Can you give an example of a reference system without fixed origin?

johnnychen94 replied:
Points in the unit sphere. It must be defined as a manifold with some atlas (coordinate charts).

juliohm commented on src/CoordinateSystemsCore/points.jl:
```diff
@@ -0,0 +1,97 @@
+end
+
+"""
+    coordinate([cs::AbstractCoordinateSystem,] x::AbstractPoint)
```
coordinates are plural in general

juliohm reviewed (commented):
I like the package name, but the design is not very clear to me. I thought we would just concentrate on coordinates and coordinates transformations first, and then leave the concept of point untouched here. Another package could depend on CoordinateSystems.jl to define a point type and implement other functionality.

johnnychen94 reviewed (commented):
When I draft this version I'm considering any points in a smooth manifold. For any local region of a manifold, there is a smooth map (chart) `ϕ` that maps from the manifold onto the euclidean space.

Thus there exist two quite different concepts:

* the point `p` in the manifold `M`. Its coordinates are very ambiguous without an atlas (the collection of charts).
* the coordinates under given chart `ϕ`: `ϕ(p)`. This is always a point in the Euclidean space. 

Thus for clarity and generality, when we speak of coordinates, we should always assume a chart on it. And that chart can be deduced from the pair of two manifolds, which constitutes the coordinate system.

So for example, when we create `Spherical{2}()`, we're actually build a coordinate system from `Spherical{2}() => Cartesian{2}()`. And it's totally doable to build a coordinate system from `Spherical{3}() => Spherical{2}() * Cartesian{1}()`, i.e., an coordinate system that maps from sphere to cylinder.

juliohm commented:
>* the point p in the manifold M. Its coordinates are very ambiguous without an atlas (the collection of charts).
>* the coordinates under given chart ϕ: ϕ(p). This is always a point in the Euclidean space.

Can we concentrate on coordinates and transformations without entering in the realm of manifolds? I think Manifolds.jl is a mature system for these more general maps and concepts. I had a more modest goal in mind when I started that issue to just provide transformations assuming everything was specified properly (charts, manifolds, atlas, ...), possibly in terms of Manifolds.jl. Basically I was after CoordinateTransformations.jl + Rotations.jl + Geodesy.jl but more organized and with a clear API. Perhaps you have a more ambitious goal in mind? I would just avoid reimplementing manifolds from scratch and consider ManifoldsBase.jl as a start.

c42f commented on src/CoordinateSystemsCore/points.jl:
```diff
@@ -0,0 +1,97 @@
+
+A point type whose origin `(0, 0, ..., 0)` is preserved under coordinate transformations.
+"""
+struct Point{CT<:FixedOriginCoordinateSystem,N,T,AT<:SVector} <: AbstractPoint{CT,N,T}
```
Should `Point` contain a field

`coord_sys::CT`

This should allow to work in the value domain more than the type domain, and opens up the possibility of coordinate systems with data when that is necessary. For example, tagged coordinate systems like you might want when working with complex rigid body data (IIUC as is done in RigidBodyDynamics.jl https://github.com/JuliaRobotics/RigidBodyDynamics.jl/blob/b9ef1d6974beff4d4fbe7dffc6dbfa65f71e0132/src/spatial/threevectors.jl#L8)

c42f commented on src/CoordinateSystemsCore/utils.jl:
```diff
@@ -0,0 +1,19 @@
+
+Apply function `f(xs...)` on coordinate system `cs`.
+
+Every point `x`∈`xs` will be maped to the working coordinate system `cs` before calling `f`, and the
```
```suggestion
Every point `x`∈`xs` will be mapped to the working coordinate system `cs` before calling `f`, and the
```

c42f commented on src/CoordinateSystemsCore/utils.jl:
```diff
@@ -0,0 +1,19 @@
+
+Apply function `f(xs...)` on coordinate system `cs`.
+
+Every point `x`∈`xs` will be maped to the working coordinate system `cs` before calling `f`, and the
+result `y` will be maped back to its original coordinate system.
```
```suggestion
Every point `x`∈`xs` will be mapped to the working coordinate system `cs` before calling `f`, and the
result `y` will be mapped back to its original coordinate system.
```

c42f commented:
Great start @johnnychen94 !

> Can we concentrate on coordinates and transformations without entering in the realm of manifolds?

I think there can be value in a package which is self-aware about differential geometry, but doesn't get caught up in the mathematical jargon and more esoteric needs of a full manifolds package in the API. It seems to me that @johnnychen94 is going in that direction?

My overall reaction to this code so far is that it's interesting and worth pursuing to see what practical functionality can be built on a `Point` which knows both its coordinate system and the particular coordinates within the chart. I think it's important that the `Point` contain the coordinate system value (often zero bytes) and not merely its type.

It's not clear to me that a coordinate system centered design is better than a transformation centered design. But having both to experiment with could be interesting and useful. (I think it's clear that Geodesy.jl has a bunch of point types which are awkward to work with...)

c42f commented:
Another thought about coordinate systems vs transformations: some transformations have a significant amount of computational work which can be cached and reused between points. Taken at face value, this favors a transformation-centric design.

`Geodesy.TransverseMercator` is an example of this (and the associated `LLAfromUTM` transformation type).  Another example is the transformations in Proj4.jl, where the underlying C library does the work of precomputing an appropriate transformation.

johnnychen94 commented:
@juliohm I think we're having only some conceptual differences but the implementations and what we're calculating are still the same. Let me explain it a bit.

The following table shows the equivalence relation between your proposal and mine.

| proposal | @juliohm | @johnnychen94 |
| ---          | ------       | --------- |
| Cartesian coordinate | `Cartesian(1, 2)` | `Point(Cartesian{2}(), (1, 2))` |
| Polar coordinate | `Polar(1, pi)` | `Point(Spherical{2}(), (1, pi))` |

The benefit of my proposal is that we're not limited to the Euclidean space, we can talk about any point in a manifold space as long as it has an atlas. I have no plan to add anything related to calculus, which is the scope of Manifolds.jl. As @c42f said, I only meant to make this a type system to represent points in generic manifold and not just in Euclidean space.

A chart is defined by the `convert_map` function, for instance, the following codes define `ϕ` and `inv(ϕ)`:

![image](https://user-images.githubusercontent.com/8684355/130311102-04295c90-77c1-42ab-bced-b4da949c836a.png)

```julia
# ϕ
function convert_map(to::Cartesian{2}, from::Spherical{2})
    @inline function (x::CoordinateType{2})
        s, c = sincos(x[2])
        return SVector(x[1] * c, x[1] * s)
    end
end
# inv(ϕ)
function convert_map(to::Spherical{2}, from::Cartesian{2})
    @inline function (x::CoordinateType{2})
        r = hypot(x[1], x[2])
        θ = atan(x[2], x[1])
        return SVector(r, θ)
    end
end
```

Transforming from one coordinate system to the other coordinate system is backed by `coordinates`:

```julia
coordinates(x::Point) = x.x
function coordinates(cs::AbstractCoordinateSystem, x::AbstractPoint)
    map_to = convert_map(cs, coordinate_system(x))
    map_to(coordinates(x))
end
```

As I tried to emphasize, the concept "coordinates" always live in a Euclidean space. And theoretically, it can be anything that is vector-like.

If we think `coordinates` as an operation to strip the coordinate system and get the raw vector representation, then `point` is kind of the inverse of `coordinates`, that it adds some
coordinate system to the raw vector representation and thus make it a point.

@juliohm So far, except for some conceptual clarification and distinction, I don't see a very distinctive difference between your proposal and mine.

---

> I think it's important that the Point contain the coordinate system value (often zero bytes) and not merely its type.

I've thought about this too:

```julia
struct CPoint{CT<:AbstractCoordinateSystem, N, T, AT} <: AbstractPoint{CT, N, T}
    chart::CT
    coordinates::AT
end
```

But I'm not sure if there will be performance differences when applying a transformation to a list of points; I was thinking of the CPU cache efficiency and compiler friendliness. I'll do more experiments to compare these two struct types and see if there are major performance differences. If there aren't then I'll merge them into one unified struct. But if there are, then some hard work is needed to bridge different point types.

---

The mystery of manifold is that a point in a given manifold can have multiple coordinates due to the choice of charts. For instance, for point `x` in a unit ball S, we can build one chart `ϕN`from the north pole and one chart `ϕS` from the south pole, and the difficulty lies in `ϕN(x) != ϕS(x)`.

I would want to also support this but it's still not clear to me how that's feasible.

c42f commented:
> But I'm not sure if there will be performance differences when applying a transformation to a list of points; I was thinking of the CPU cache efficiency and compiler friendliness. I'll do more experiments to compare these two struct types and see if there are major performance differences.

My guess is that `sizeof(chart) == 0` in many practical cases so performance and struct layout should be identical.  Cases where the size is nonzero indicate that there's something parametric about `chart` that the user needs to track as data rather than in the type. IIUC, https://github.com/JuliaRobotics/RigidBodyDynamics.jl/blob/b9ef1d6974beff4d4fbe7dffc6dbfa65f71e0132/src/spatial/threevectors.jl#L8 is exactly one such case. If only the type of the chart is included, CoordinateSystems.jl won't be usable in those cases.

c42f commented:
> The mystery of manifold is that a point in a given manifold can have multiple coordinates due to the choice of charts. For instance, for point `x` in a unit ball S, we can build one chart `ϕN`from the north pole and one chart `ϕS` from the south pole, and the difficulty lies in `ϕN(x) != ϕS(x)`.
> 
> I would want to also support this but it's still not clear to me how that's feasible.

But I think you've already implemented this for one example: Normal Spherical coordinates and Cartesian coordinates are just two different coordinate charts for 3D Euclidean space, and your `convert_map` provides the mapping between the coordinates in those two charts?

johnnychen94 committed:
rewrite to resolve given comments

change of behaviors:

* `Point` now has an extra field `chart` to keep track of its coordinate system.
* `with_coordinate_system(chart, x)` doesn't maps back to `x`'s original
  coordinate system. If everything works well, the transition of
  of coordinate systems should be automatically and lazily.
* `coordinates(x, y)` get the local coordinates of `y` under `x`'s coordinate
  system.

enhancement:

* coordinates is expected to be vector-like, it doesn't need to be
  `SVector`.
* implement generic N-dimensional transition between Cartesian and Spherical
* add documentation

rename and maintenance:

* seperate a sub-package `CoordinateSystemsCore` where it focus only
  on coordinates, points, and coordinate transitions.
* `AbstractCoordinateSystem` => `CoordinateSystem`
* `AbstractCoordinateTransformation` => `CoordinateTransformation`
* for coordinate system: `cs` => `chart`
* `convert_map` => `transition_map`

johnnychen94 reviewed (commented):
d4f6f05e3c3f733dca9c63234886be933dffc2ad is a big commit, you can check the summary for some information. Here I'd like to resolve some of the previous comments.

> Can you explain how we benefit from a Core submodule?

My idea is to let the Core submodule become independent as possible as it can, and only cover necessary traits, utils, transformations and coordinates so that other packages can work on top of it.

People might have different ideas on whether we should attach a coordinate to transformations or other functions. It doesn't matter. By separating a core module I want to support both packages that lives in typical euclidean world and packages that have attached coordinate systems.

In the latest commit d4f6f05e3c3f733dca9c63234886be933dffc2ad I've made this an independent package so if you don't like how I build Transformations, you can start your own version of it.

> I would remove the Abstract prefix from most types to make them short. Also, maybe we could leave the Point type outside this package and only deal with the coordinate transformations. The point itself can be defined in separate packages that consume coordinates.

I've removed the Abstract prefix for `AbstractCoordinateTransformation` and `AbstractCoordinateSystem`. I still keep the `AbstractPoint` prefix because we're using `Point` as a concrete type here.

I hope I've made it clear in https://github.com/johnnychen94/CoordinateSystems.jl/pull/1#issuecomment-903062046 that there is probably only one `Point` type here and it's closely connected with the coordinate systems.

> I would not fall back to SVector or NTuple because that is gonna cause the same issues we have currently with SVector everywhere. We need to really enforce some Cartesian type to avoid SVector proliferation.

In commit d4f6f05e3c3f733dca9c63234886be933dffc2ad I've taken efforts to allow `AbstractVector`, this makes it possible to deal with very high dimensional spaces. (Although I haven't had a clear view of how it would work in practice)

> Can you give an example of a reference system without fixed origin?

`FixedOriginCoordinateSystem` is removed.

> Should Point contain a field

I've made `chart` as a field of `Point`. The name "chart" is used to differentiate with the function name `coordinate_system`.

---

After this, there are several things I'm planning to experiment with:

- if making `AbstractPoint <: AbstractVector` make things easier to work with
- define some promote rules so that when different coordinate systems meet together, they know which coordinate they're supposed to promote to
- mixed dimension coordinate transitions (e.g., `Cartesian{2}() => Cartesian{3}()` and vice versa)

Depending on how big these changes are, they might be included in a separate PR, though.

johnnychen94 committed:
fix CI setup and old version compatibility

c42f commented on CoordinateSystemsCore/src/points.jl:
```diff
@@ -0,0 +1,148 @@
+@inline Base.ndims(::AbstractPoint{CT,N}) where {CT,N} = N
+@inline Base.:(==)(x::AbstractPoint, y::AbstractPoint) = with_coordinate_system(==, coordinate_system(x), x, y)
+@inline function Base.isapprox(x::AbstractPoint, y::AbstractPoint; kwargs...)
+    with_coordinate_system(coordinate_system(x), x, y) do p, q
```
It strikes me that `with_coordinate_system` is very like `Base.promote` in function, if not in API. Perhaps that analogy is worth exploring?

johnnychen94 replied:
Yes! I had the same feeling, and that's why I'm thinking of making `AbstractPoint <: AbstractVector`, and then maybe it's possible to hide all the coordinate chart transition details under the scene while still reason as if we're on the Euclidean coordinates.

c42f replied:
It could make sense to adopt a `promote`-like API (ie, return `p` and `q` rather than passing them to a callback) to make the use of this more imperative in style?

About `AbstractPoint <: AbstractVector`... maybe? I'm not sure what that implies!

c42f commented on CoordinateSystemsCore/src/transition_map.jl:
```diff
@@ -0,0 +1,85 @@
+@inline transition_map(to::CT, from::CT) where CT<:CoordinateSystem = identity
+
+@inline function transition_map(to::Cartesian{N}, from::Spherical{N}) where N
+    Base.@propagate_inbounds function Spherical_to_Cartesian(x::AbstractVector{T}) where T
```
A minor thing - I think it would be clearer here if `Spherical_to_Cartesian` were just pulled out into a normal function (rather than being a closure). Then return that from `transition_map`.

johnnychen94 replied:
Agreed, I've thought about this too and maybe even make CoordinateSystem a functor.
This would add more additional mess to this PR so I might work it in the next PR.

c42f commented on test/transformations.jl:
```diff
@@ -0,0 +1,46 @@
+            q = @inferred trans(p)
+            @test coordinates(p, q) == [4, 6]
+
+            trans = Translation(Spherical{2}(), (0, pi)) # becomes a rotation
```
I find this *super* confusing! I would expect `Translation` to always be a translation in the Cartesian system.

It's true that you can get a subset of rotations with this interpretation of "translation", but you can't get all of them.
